test/logproxy/logproxy.go

```diff
@@ -47,79 +56,330 @@ func (p *Proxy) Run() error {
 	}
+}
+
+func (p *Proxy) serv(netconn net.Conn) error {
+	conn, err := net.Dial("tcp", fmt.Sprintf("%s:%s", p.toHost, p.toPort))
+	if err != nil {
+		return err
+	}
+
+	frontend := pgproto3.NewFrontend(bufio.NewReader(conn), conn)
+	cl := pgproto3.NewBackend(bufio.NewReader(netconn), netconn)
+
+	//handle startup messages
+	if err = startup(netconn, frontend, cl); err != nil {
+		return err
+	}
+
+	defer p.Flush()
 
 	for {
 		msg, err := cl.Receive()
+		if err != nil {
+			return fmt.Errorf("failed to receive msg from client %w", err)
+		}
 
-		cb(msg)
-
+		//writing request data to buffer
+		byt, err := encodeMessage(msg)
 		if err != nil {
-			return fmt.Errorf(failedToReceiveMessage, err)
+			return fmt.Errorf("failed to convert %w", err)
+		}
+		p.interceptedData = append(p.interceptedData, byt...)
+		if len(p.interceptedData) > 1000000 {
+			spqrlog.Zero.Debug().Msg("flushing buffer")
+			err = p.Flush()
+			if err != nil {
+				return fmt.Errorf("failed to write to file %w", err)
+			}
 		}
+
+		//send to frontend
 		frontend.Send(msg)
 		if err := frontend.Flush(); err != nil {
-			return fmt.Errorf(failedToReceiveMessage, err)
+			return fmt.Errorf("failed to send msg to bd %w", err)
 		}
-		for {
-			retmsg, err := frontend.Receive()
+
+		proc := func(msg pgproto3.BackendMessage) error {
+			cl.Send(msg)
+			if err := cl.Flush(); err != nil {
+				return fmt.Errorf("failed to send msg to client %w", err)
+			}
+			return nil
+		}
+		err = recieveBackend(frontend, proc)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (p *Proxy) Flush() error {
+	spqrlog.Zero.Debug().Msg("flush")
+	f, err := os.OpenFile(p.logFileName, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	_, err = f.Write(p.interceptedData)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func parseFile(f *os.File) (time.Time, pgproto3.FrontendMessage, error) {
+	// 15 byte - timestamp
+	// 1 byte - message header
+	// 4 bytes - message length (except header)
+	// ?? bytes - message bytes
+
+	//timestamp
+	timeb := make([]byte, 15)
+	_, err := f.Read(timeb) //err
+	if err != nil {
+		return time.Now(), nil, err
+	}
+	var ti time.Time
+	err = ti.UnmarshalBinary(timeb)
+	if err != nil {
+		return time.Now(), nil, err
+	}
+
+	//header
+	tip := make([]byte, 1)
+
+	_, err = f.Read(tip)
+	if err != nil {
+		return time.Now(), nil, err
+	}
+	if string(tip) == "X" {
+		return time.Now(), nil, io.EOF
+	}
+
+	//size
+	rawSize := make([]byte, 4)
+
+	_, err = f.Read(rawSize)
+	if err != nil {
+		return time.Now(), nil, err
+	}
+
+	msgSize := int(binary.BigEndian.Uint32(rawSize) - 4)
+
+	//message
+	msg := make([]byte, msgSize)
+	_, err = f.Read(msg)
+	if err != nil {
+		return time.Now(), nil, err
+	}
+
+	fm := &pgproto3.Query{}
+	err = fm.Decode(msg)
+	if err != nil {
+		return time.Now(), nil, err
+	}
+
+	return ti, fm, nil
+}
+
+func startup(netconn net.Conn, frontend *pgproto3.Frontend, cl *pgproto3.Backend) error {
```

Comment:
there is same exact function in router/client
we should reuse code, not write it twice